src/type-data-frame.c

```diff
@@ -59,21 +59,48 @@ SEXP vctrs_new_data_frame(SEXP args) {
     Rf_errorcall(R_NilValue, "`x` must be a list");
   }
 
-  R_len_t size = df_size_from_list(x, n);
+  bool has_n = !Rf_isNull(n);
+  bool has_names = !Rf_isNull(r_names(x));
+
+  R_len_t size = -1;
+
+  SEXP out = PROTECT(r_maybe_duplicate(x));
+
+  for (SEXP attr_in = attrib; attr_in != R_NilValue; attr_in = CDR(attr_in)) {
+    SEXP tag = TAG(attr_in);
+
+    if (tag == R_NamesSymbol) {
+      // names are ignored if the input is named
+      if (has_names) continue;
+    } else if (tag == R_RowNamesSymbol) {
+      // row.names is ignored if n is provided
+      if (has_n) continue;
```

Comment:
It looks like this is not handling character rownames correctly. They should be assigned in the attributes list.

Reply:
The tests suggest it works.

Reply:
I don't understand how it works.

Reply:
We copy over the `"row.names"` attribute only if `n` is not provided. If the caller passes `n`, it takes precedence over `row.names` in the dots.

Reply:
This is not correct. We should assign row names if they are character.

Reply:
What if `n` and `"row.names"` disagree? Why is this important?

It's easy to *not* pass `n`. It's more difficult, and slower, to avoid passing `"row.names"`.

Reply:
It is important to have a good and predictable UI.

If `n` and `row.names` disagree, please make it an error. This is a fast check.